File: 0076-minimum-window-substring/solution.cpp
#include <string>
#include <unordered_map>
#include <climits>
using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        if (t.empty()) {
            return "";
        }

        unordered_map<char, int> need;
        for (char c : t) {
            need[c]++;
        }

        unordered_map<char, int> have;
        int formed = 0;
        int required = (int)need.size();
        int left = 0;
        int bestLen = INT_MAX;
        int bestLeft = 0;

        for (int right = 0; right < (int)s.size(); right++) {
            char c = s[right];
            have[c]++;

            if (need.count(c) && have[c] == need[c]) {
                formed++;
            }

            while (formed == required) {
                if (right - left + 1 < bestLen) {
                    bestLen = right - left + 1;
                    bestLeft = left;
                }

                char leftChar = s[left];
                have[leftChar]--;
                if (need.count(leftChar) && have[leftChar] < need[leftChar]) {
                    formed--;
                }
                left++;
            }
        }

        return bestLen == INT_MAX ? "" : s.substr(bestLeft, bestLen);
    }
};

File: 0076-minimum-window-substring/solution.py
class Solution:
    def minWindow(self, s, t):
        if not t:
            return ''

        need = {}
        for char in t:
            need[char] = need.get(char, 0) + 1

        have = {}
        formed = 0
        required = len(need)
        left = 0
        best = (float('inf'), 0, 0)

        for right, char in enumerate(s):
            have[char] = have.get(char, 0) + 1
            if char in need and have[char] == need[char]:
                formed += 1

            while formed == required:
                if right - left + 1 < best[0]:
                    best = (right - left + 1, left, right)

                left_char = s[left]
                have[left_char] -= 1
                if left_char in need and have[left_char] < need[left_char]:
                    formed -= 1
                left += 1

        return '' if best[0] == float('inf') else s[best[1]:best[2] + 1]